src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/NeuralNetInference.java

```diff
@@ -0,0 +1,291 @@
+                    getVariantInfoString(variant),
+                    variant.isSNP() ? "SNP" : variant.isIndel() ? "INDEL" : "OTHER");
+
+            if (curBatchSize == transferBatchSize) {
+                if (waitforBatchCompletion == true) {
+                    // wait for the last batch to complete before we start a new one
+                    asyncWriter.waitForPreviousBatchCompletion(1, TimeUnit.MINUTES);
+                    waitforBatchCompletion = false;
+                    pythonExecutor.getAccumulatedOutput();
+                }
+                executePythonCommand();
+                waitforBatchCompletion = true;
+                curBatchSize = 0;
+                batchList = new ArrayList<>(transferBatchSize);
+            }
+
+            batchList.add(outDat);
+            curBatchSize++;
+        } catch (UnsupportedEncodingException e) {
+            throw new GATKException("Trying to make string from reference, but unsupported encoding UTF-8.", e);
+        }
+
+    }
+
+    private String getVariantDataString(final VariantContext variant){
+        return String.format("%s\t%d\t%s\t%s",
+                variant.getContig(),
+                variant.getStart(),
+                variant.getReference().getBaseString(),
+                variant.getAlternateAlleles().toString()
+        );
+
+    }
+
+    private String getVariantInfoString(final VariantContext variant){
+        // Create a string that will easily be parsed as a python dictionary
+        String varInfo = "";
+        for (final String attributeKey : variant.getAttributes().keySet()) {
+            varInfo += attributeKey + "=" + variant.getAttribute(attributeKey).toString().replace(" ", "").replace("[", "").replace("]", "") + ";";
+        }
+        return varInfo;
+    }
+
+    @Override
+    public Object onTraversalSuccess() {
+        if (waitforBatchCompletion) {
+            asyncWriter.waitForPreviousBatchCompletion(1, TimeUnit.MINUTES);
+            pythonExecutor.getAccumulatedOutput();
+        }
+        if (curBatchSize > 0){
+            executePythonCommand();
+            asyncWriter.waitForPreviousBatchCompletion(1, TimeUnit.MINUTES);
+            pythonExecutor.getAccumulatedOutput();
+        }
+
+        pythonExecutor.sendSynchronousCommand("tempFile.close()" + NL);
+        pythonExecutor.sendSynchronousCommand("fifoFile.close()" + NL);
+        pythonExecutor.terminate();
+
+        writeOutputVCFWithScores();
+
+        return true;
+    }
+
+    private void executePythonCommand(){
+        final String pythonCommand = String.format(
+                "vqsr_cnn.score_and_write_batch(model, tempFile, fifoFile, %d, %d)", curBatchSize, inferenceBatchSize) + NL;
+        pythonExecutor.sendAsynchronousCommand(pythonCommand);
+        asyncWriter.startAsynchronousBatchWrite(batchList);
+    }
+
+
+    private void writeOutputVCFWithScores(){
+        writeVCFHeader();
+        try (Scanner scoreScan = new Scanner(scoreFile)) {
```

Comment:
Can you remove the class level vcfWriter variable, and localize it to try-with-resources:

```
        try (final Scanner scoreScan = new Scanner(scoreFile);
             final VariantContextWriter vcfWriter = createVCFWriter(new File(outputFile))) {
            scoreScan.useDelimiter("\\n");
            writeVCFHeader(vcfWriter); // or call this getOutputHeader, have it return the header, and write it here
```
Then all the closing is handled automatically (you can remove the explicit vcfWriter.close() below), and the resource handling will be nicely symmetric.

Reply:
Nice cleanup, fixed